src/WebJobs.Extensions.DocumentDB/Bindings/DocumentDBEnumerableValueBinder.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DocumentDB
+{
+    internal class DocumentDBEnumerableValueBinder<T> : IValueBinder
+    {
+        private DocumentDBContext _context;
+
+        public DocumentDBEnumerableValueBinder(DocumentDBContext context)
+        {
+            _context = context;
+        }
+
+        public Type Type
+        {
+            get
+            {
+                return typeof(IEnumerable<T>);
+            }
+        }
+
+        public object GetValue()
+        {
+            Uri collectionUri = UriFactory.CreateDocumentCollectionUri(_context.ResolvedAttribute.DatabaseName, _context.ResolvedAttribute.CollectionName);
+
+            List<Document> finalResults = new List<Document>();
+
+            string continuation = null;
+
+            do
+            {
+                DocumentQueryResponse response = DocumentDBUtility.RetryAsync(
+                    () => _context.Service.ExecuteNextAsync(collectionUri, _context.ResolvedAttribute.SqlQuery, continuation),
+                    _context.MaxThrottleRetries, codesToIgnore: HttpStatusCode.NotFound).Result;
+
+                if (response == null)
+                {
+                    return null;
+                }
+
+                finalResults.AddRange(response.Results);
+                continuation = response.ResponseContinuation;
+            }
+            while (!string.IsNullOrEmpty(continuation));
```

Comment:
It might be worthwhile to add some tracing around this / enhanced errors.  

For script languages (which include binding time in overall timeout time) we could probably decrease confusion if, instead of throwing a generic timeout error, we indicate that the timeout occurred due to the query and not user code.

Not in the scope of this PR, think it's more of an invoker issue